File: is-palindrome-permutation/solution1.cpp
#include <cctype>
#include <string>
#include <iostream>
#include <unordered_map>

bool is_permutation(std::string&& str) {
    if (str.empty()) return true;
    std::unordered_map<int, int> count;
    for (auto& ch : str) {
        if (ch != ' ') {
            count[std::tolower((unsigned char)ch)]++;
        }
    }
    bool odd_already_exists = false;
    for (const auto& cnt : count) {
        if (cnt.second % 2 != 0) {
            if (odd_already_exists) return false;
            odd_already_exists = true;
        }
    }
    return true;
}

int main() {
    std::cout << "Tact Coa " << std::boolalpha << is_permutation("Tact Coa") << std::endl;
    std::cout << "Tact Coa " << std::boolalpha << is_permutation("Tact Coap") << std::endl;
}

File: is-palindrome-permutation/solution1.js
const is_permutation = (str) => {
    if (typeof str != 'string') return false;
    if (!str.length) {
        return true;
    }
    let count = new Map();
    str = str.toLowerCase();
    for (let ch of str) {
        if (ch != ' ') {
            count.set(ch, ((count.get(ch) || 0) + 1));
        }
    }
    let odd_exists = false;
    for (let [ch, cnt] of count) {
        if (cnt % 2 != 0) {
            if (odd_exists) return false;
            odd_exists = true;
        }
    }
    return true;
}

console.log("Tact Coa: ", is_permutation("Tact Coa"));
console.log("Tact Coap: ", is_permutation("Tact Coap"));
